nixos/tests/gitlab-fetch-private.nix

```diff
@@ -0,0 +1,371 @@
+          domain = "${domain}";
+          owner = "alice";
+          repo = "test-alice";
+          rev = "refs/heads/master";
+          sha256 = "sha256-HqozwmyEGzXUlSL6zgInlycalxROnD2QOSPRwkrw8sQ=";
+        }
+      '';
+
+      # The fetchers don't honor the OS trusted CAs as this would be impure.
+      # To make it accept our custom (self-signed) CA, it must be added to cacert.
+      fetchWithRepoScope = pkgs.writeText "fetch-with-repo-scope.nix" ''
+        let
+          pkgs = import <nixpkgs> {};
+          fetchFromGitLab = pkgs.fetchFromGitLab.override {
+            fetchgit = pkgs.fetchgit.override {
+              cacert = ${cacert_with_extras};
+            };
+          };
+        in pkgs.invalidateFetcherByDrvHash fetchFromGitLab {
+          domain = "${domain}";
+          private = true;
+          # varPrefix: default, no prefix
+          owner = "alice";
+          repo = "test-alice";
+          rev = "refs/heads/master";
+          sha256 = "sha256-HqozwmyEGzXUlSL6zgInlycalxROnD2QOSPRwkrw8sQ=";
+          forceFetchGit = true;
+        }
+      '';
+
+      # The fetchers don't honor the OS trusted CAs as this would be impure.
+      # To make it accept our custom (self-signed) CA, it must be added to cacert.
+      fetchWithApiScope = pkgs.writeText "fetch-with-api-scope.nix" ''
+        let
+          pkgs = import <nixpkgs> {};
+          fetchFromGitLab = pkgs.fetchFromGitLab.override {
+            fetchzip = pkgs.fetchzip.override {
+              fetchurl = pkgs.fetchurl.override {
+                cacert = ${cacert_with_extras};
+              };
+            };
+          };
+        in pkgs.invalidateFetcherByDrvHash fetchFromGitLab {
+          domain = "${domain}";
+          private = true;
+          varPrefix = "MYPREFIX";
+          owner = "alice";
+          repo = "test-alice";
+          rev = "refs/heads/master";
+          sha256 = "sha256-XHNNJr5NrfLMnom2ZQkkhmZ/7q8eOc563N4AbyF3K84=";
+        }
+      '';
+
+      # Wait for all GitLab services to be fully started.
+      waitForServices = ''
+        gitlab.wait_for_unit("gitaly.service")
+        gitlab.wait_for_unit("gitlab-workhorse.service")
+        # https://github.com/NixOS/nixpkgs/issues/132295
+        # gitlab.wait_for_unit("gitlab-pages.service")
+        gitlab.wait_for_unit("gitlab-mailroom.service")
+        gitlab.wait_for_unit("gitlab.service")
+        gitlab.wait_for_unit("gitlab-sidekiq.service")
+        gitlab.wait_for_file("${nodes.gitlab.config.services.gitlab.statePath}/tmp/sockets/gitlab.socket")
+        gitlab.wait_until_succeeds("curl -sSf http://gitlab/users/sign_in")
+      '';
+
+      # The actual test of GitLab.
+      test = ''
+        gitlab.succeed(
+            "curl -isSf http://gitlab | grep -i location | grep http://gitlab/users/sign_in"
+        )
+        gitlab.succeed(
+            "${pkgs.sudo}/bin/sudo -u gitlab -H gitlab-rake gitlab:check 1>&2"
+        )
+        gitlab.succeed(
```

Comment:
I would prefer for the client to be in a separate node. It's just slightly more true to real use.